src/NuGet.Core/NuGet.Packaging/Rules/DependeciesGroupsForEachTFMRule.cs

```diff
@@ -0,0 +1,176 @@
+        }
+        public IEnumerable<PackagingLogMessage> Validate(PackageArchiveReader builder)
+        {
+            var files = builder.GetFiles().
+                Where(t => PathUtility.GetPathWithDirectorySeparator(t).Count(m => m == Path.DirectorySeparatorChar) > 1);
+            var packageNuspec = builder.GetNuspec();
+            return Validate(files, packageNuspec);
+        }
+
+        internal IEnumerable<PackagingLogMessage> Validate(IEnumerable<string> files, Stream packageNuspec)
+        {
+            var managedCodeConventions = new ManagedCodeConventions(new RuntimeGraph());
+            var isCompatible = managedCodeConventions.Properties["tfm"].CompatibilityTest;
+            var collection = new ContentItemCollection();
+            collection.Load(files);
+
+            var libItems = ContentExtractor.GetContentForPattern(collection, managedCodeConventions.Patterns.CompileLibAssemblies);
+            var refItems = ContentExtractor.GetContentForPattern(collection, managedCodeConventions.Patterns.CompileRefAssemblies);
+
+            var tfmsFromFilesSet = new HashSet<NuGetFramework>();
+            tfmsFromFilesSet.AddRange(ContentExtractor.GetGroupFrameworks(libItems).ToArray());
+            tfmsFromFilesSet.AddRange(ContentExtractor.GetGroupFrameworks(refItems).ToArray());
+            var tfmsFromFiles = tfmsFromFilesSet.Where(t => t != NuGetFramework.Parse("net00")).ToList();
+
+            List<NuGetFramework> tfmsFromNuspec = new List<NuGetFramework>();
+            var stream = new NuspecReader(packageNuspec);
+            var nuspec = stream.Xml;
+            if (nuspec != null)
+            {
+                string name = nuspec.Root.Name.Namespace.ToString();
+                tfmsFromNuspec = nuspec.Descendants(XName.Get("dependencies", name)).Elements().
+                    Attributes("targetFramework").Select(f => NuGetFramework.Parse(f.Value)).ToList();
+            }
+
+            if (tfmsFromFiles.Count != 0 && tfmsFromNuspec.Count != 0)
+            {
+                foreach(var fileTFM in tfmsFromFiles)
+                {
+                    foreach(var nuspecTFM in tfmsFromNuspec)
+                    {
+                        if (!tfmsFromNuspec.Contains(fileTFM))
+                        {
+                            _noExactMatchesFromFile.Add(fileTFM);
+                        }
+
+                        if (!tfmsFromFiles.Contains(nuspecTFM))
+                        {
+                            _noExactMatchesFromNuspec.Add(nuspecTFM);
+
+                            if (isCompatible(fileTFM, nuspecTFM))
+                            {
+                                _compatNotExactMatches.Add(fileTFM);
+                            }
+                        }
+                    }
+                }
+
+                _noExactMatchesFromFile.RemoveWhere(IsCompat);
+                _noExactMatchesFromNuspec.RemoveWhere(IsCompat);
```

Comment:
Do we want to throw two warnings for the same TFM?

Reply:
I see three scenarios:
* Asset group don't have exact match dependency group, but there is a compatible dependency group
* Asset group don't have an exact match dependency group, and there is no compatible dependency group
* dependency group does not have an exact match asset group

Using these three categories, it's not possible for a single TFM to be in more than one category.

Reply:
That's why i remove the ones that fall into the second one. Because the current logic is that a TFM has no exact match AND has a compat match. And that's how they are grouped.